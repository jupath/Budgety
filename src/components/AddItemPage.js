import React from 'react';

const AddItemPage = () => (
  <div>
    AddItemPage
  </div>
);

export default AddItemPage;
