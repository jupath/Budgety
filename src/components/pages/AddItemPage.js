import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { startAddItem } from '../../actions/budget';
import AddItemForm from '../forms/AddItemForm';

export class AddItemPage extends Component {
  state = {
    itemTypeState: undefined,
  };

  componentWillMount() {
    this.checkUrlParams();
  }

  checkUrlParams = () => {
    const { itemType } = this.props;
    if (itemType === 'income') {
      this.setState({ itemTypeState: 'income' });
    } else if (itemType === 'expense') {
      this.setState({ itemTypeState: 'expense' });
    } else {
      this.props.history.push('/');
    }
  };

  handleAddItem = (item) => {
    this.props.addItem(item);
    this.props.history.push('/');
  };

  render() {
    return (
      <div className="container px-lg-7">
        <h2 className="pt-4">Add new {this.state.itemTypeState}</h2>
        <AddItemForm handleAddItem={this.handleAddItem} itemType={this.state.itemTypeState} />
      </div>
    );
  }
}

AddItemPage.propTypes = {
  addItem: PropTypes.func.isRequired,
  history: PropTypes.object.isRequired,
  itemType: PropTypes.string.isRequired,
};

const mapStateToProps = (state, props) => ({
  itemType: props.match.params.type,
});

const mapDispatchToProps = dispatch => ({
  addItem: item => dispatch(startAddItem(item)),
});

export default connect(mapStateToProps, mapDispatchToProps)(AddItemPage);
