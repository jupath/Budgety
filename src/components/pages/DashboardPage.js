import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import BudgetSummary from '../layout/BudgetSummary';
import ListFilterForm from '../forms/ListFilterForm';
import ListHeader from '../list/ListHeader';
import BudgetList from '../list/BudgetList';
import itemsSelector from '../../selectors/items';

export const DashboardPage = props => (
  <div>
    <BudgetSummary />
    <ListFilterForm />
    <div className="container">
      <div className="row">
        <div className="col-sm-6">
          <ListHeader listName="Expenses" itemsType="expense" />
          {props.expenses.length > 0 && <BudgetList items={props.expenses} />}
        </div>
        <div className="col-sm-6">
          <ListHeader listName="Incomes" itemsType="income" />
          {props.incomes.length > 0 && <BudgetList items={props.incomes} />}
        </div>
      </div>
    </div>
  </div>
);

const mapStateToProps = state => ({
  incomes: itemsSelector(state.items, 'income', state.filters),
  expenses: itemsSelector(state.items, 'expense', state.filters),
});

DashboardPage.propTypes = {
  incomes: PropTypes.arrayOf(PropTypes.object),
  expenses: PropTypes.arrayOf(PropTypes.object),
};

DashboardPage.defaultProps = {
  incomes: [],
  expenses: [],
};

export default connect(mapStateToProps)(DashboardPage);
