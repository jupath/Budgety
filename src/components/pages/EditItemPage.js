import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Button, Modal, ModalHeader, ModalBody, ModalFooter } from 'reactstrap';
import { startUpdateItem, startDeleteItem } from '../../actions/budget';
import AddItemForm from '../forms/AddItemForm';

export class EditItemPage extends Component {
  state = {
    modal: false,
  }

  onDeleteItem = () => {
    this.props.deleteItem(this.props.item.id);
    this.toggleModal();
    this.props.history.push('/');
  };

  toggleModal = () => {
    this.setState({
      modal: !this.state.modal,
    });
  }

  handleUpdateItem = (item) => {
    this.props.updateItem(this.props.item.id, item);
    this.props.history.push('/');
  }

  render() {
    return (
      <div className="container mt-4 px-lg-7">
        <h2>Update {this.props.item.itemType}</h2>
        <AddItemForm handleAddItem={this.handleUpdateItem} item={this.props.item} />
        <Button className="mt-2 px-4 py-2 text-uppercase" id="toggle-modal" color="danger" onClick={this.toggleModal}>Delete</Button>
        <Modal isOpen={this.state.modal} toggle={this.toggleModal}>
          <ModalHeader toggle={this.toggleModal}>Delete item</ModalHeader>
          <ModalBody>
            Are you sure you want to delete this?
          </ModalBody>
          <ModalFooter>
            <Button color="danger" id="delete" onClick={this.onDeleteItem}>Delete</Button>{' '}
            <Button color="secondary" id="cancel" onClick={this.toggleModal}>Cancel</Button>
          </ModalFooter>
        </Modal>
      </div>
    );
  }
}

EditItemPage.propTypes = {
  updateItem: PropTypes.func.isRequired,
  deleteItem: PropTypes.func.isRequired,
  item: PropTypes.objectOf(PropTypes.any).isRequired,
  history: PropTypes.object,
};

const mapStateToProps = (state, props) => ({
  item: state.items.find(item => item.id === props.match.params.id),
});

const mapDispatchToState = dispatch => ({
  updateItem: (id, item) => dispatch(startUpdateItem(id, item)),
  deleteItem: id => dispatch(startDeleteItem(id)),
});

export default connect(mapStateToProps, mapDispatchToState)(EditItemPage);
