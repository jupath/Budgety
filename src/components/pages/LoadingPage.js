import React from 'react';
import loader from '../../images/loading.svg';

const LoadingPage = () => (
  <div className="loader">
    <img src={loader} alt="loading..." />
  </div>
);

export default LoadingPage;
