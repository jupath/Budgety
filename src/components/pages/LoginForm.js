import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Button, Form, FormGroup, Label, Input } from 'reactstrap';
import { startSignInWithEmailAndPassword } from '../../actions/auth';

class LoginForm extends Component {
  state = {
    error: undefined,
  }

  doSignIn = (email, password) => {
    this.props.signInUser(email, password)
      .then(() => {
        if (this.props.authError) {
          this.setState({ error: this.props.authError });
        } else {
          this.setState({ error: undefined });
        }
      });
  }

  handleSignIn = (event) => {
    event.preventDefault();
    const email = event.target.elements.email.value;
    const password = event.target.elements.password.value;

    if (email === '' || password === '') {
      this.setState({ error: 'Please, fill out every field!' });
    } else {
      this.doSignIn(email, password);
    }
  }

  render() {
    return (
      <div>
        <Form onSubmit={this.handleSignIn}>
          <FormGroup>
            <Label className="sr-only" for="email">Email</Label>
            <Input
              type="email"
              name="email"
              id="email"
              placeholder="Email"
              required="required"
            />
          </FormGroup>
          <FormGroup>
            <Label className="sr-only" for="password">Password</Label>
            <Input
              type="password"
              name="password"
              placeholder="Password"
              required="required"
            />
          </FormGroup>
          { this.state.error && <p className="text-danger">{this.state.error}</p> }
          <Button color="success">Login</Button>
        </Form>
      </div>
    );
  }
}

LoginForm.propTypes = {
  signInUser: PropTypes.func.isRequired,
  authError: PropTypes.string,
};

LoginForm.defaultProps = {
  authError: undefined,
};

const mapStateToProps = state => ({
  authError: state.auth.error,
});

const mapDispatchToProps = dispatch => ({
  signInUser: (email, password) => dispatch(startSignInWithEmailAndPassword(email, password)),
});

export default connect(mapStateToProps, mapDispatchToProps)(LoginForm);
