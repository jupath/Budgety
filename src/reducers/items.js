export default (state = [], action) => {
  switch (action.type) {
    case 'ADD_ITEM':
      return [...state, action.item];
    case 'UPDATE_ITEM':
      return state.map((item) => {
        if (item.id === action.id) {
          return {
            ...item,
            ...action.item,
          };
        }
        return item;
      });
    case 'DELETE_ITEM':
      return state.filter(item => item.id !== action.id);
    default:
      return state;
  }
};
