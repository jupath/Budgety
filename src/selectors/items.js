import moment from 'moment';

export default (items, itemsType, { startDate, endDate, sortBy }) => items.filter((item) => {
  const date = moment(item.date);
  const startDateMatch = startDate ? startDate.isSameOrBefore(date, 'day') : true;
  const endDateMatch = endDate ? endDate.isSameOrAfter(date, 'day') : true;
  const itemTypeMatch = item.itemType === itemsType;

  return startDateMatch && endDateMatch && itemTypeMatch;
}).sort((a, b) => {
  if (sortBy === 'date') {
    return a.date < b.date ? 1 : -1;
  } else if (sortBy === 'value') {
    return a.amount < b.amount ? 1 : -1;
  }
});
