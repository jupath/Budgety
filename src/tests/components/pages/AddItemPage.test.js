import React from 'react';
import { shallow } from 'enzyme';
import { AddItemPage } from '../../../components/pages/AddItemPage';
import items from '../../fixtures/items';

let addItem, history, itemType, wrapper;

beforeEach(() => {
  addItem = jest.fn();
  history = { push: jest.fn() };
  itemType = 'income';
  wrapper = shallow(<AddItemPage addItem={addItem} itemType={itemType} history={history} />);
});

test('should render AddItemPage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle handleAddItem correctly', () => {
  wrapper.find('AddItemForm').prop('handleAddItem')(items[1]);
  expect(addItem).toHaveBeenLastCalledWith(items[1]);
  expect(history.push).toHaveBeenLastCalledWith('/');
});
