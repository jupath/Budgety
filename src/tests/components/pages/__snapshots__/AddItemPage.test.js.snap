// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddItemPage correctly 1`] = `
<div
  className="container px-lg-7"
>
  <h2
    className="pt-4"
  >
    Add new 
    income
  </h2>
  <AddItemForm
    handleAddItem={[Function]}
    item={Object {}}
    itemType="income"
  />
</div>
`;
